### PR TITLE
Add file path validation

### PR DESCRIPTION
**Pull request by mreishus**
## Test plan (Default settings)

Let's start the server and check three different types of requests:

* Start server `npm start`
* (A) `curl -i 'http://127.0.0.1:4747/file?path=tests/bootstrap.css'`
  * Allowed path, exists
* (B) `curl -i 'http://127.0.0.1:4747/file?path=tests/bootstrap2.css'`
  * Allowed path, does not exist
* (C) `curl -i 'http://127.0.0.1:4747/file?path=/etc/resolv.conf` 
  * Disallowed path, exists

*Expect to see:* 

(B) and (C) should look the same to the user: 404. A user should not be able to see slight differences between (B) and (C) to see which files exists where on the filesystem.

*Env variable*:

The new env variable should work. (See below)

### Create test files

```bash
mkdir /tmp/minifytest
```

```bash
cat << EOF > /tmp/minifytest/sample.js
// This is a sample JavaScript file for minification test
console.log('Hello, world!'); // Logs 'Hello, world!' to the console
console.log('This is a test.'); // Logs 'This is a test.' to the console

function testFunc() {
    var testVar = 'This is a test variable';
    return testVar;
}
EOF
```

```bash
cat << EOF > /tmp/minifytest/sample.css
/* This is a sample CSS file for minification test */
body {
    background-color: #f00; /* Sets the background color to red */
}

h1 {
    color: #00f; /* Sets the color of h1 elements to blue */
    margin: 0 auto; /* Centers h1 elements */
}
EOF
```

### Set correct base path and expect to access these files

```bash
export MINIFIERS_BASE_PATH=/tmp/minifytest
npm start
```

```bash
curl -i 'http://localhost:4747/file?path=/tmp/minifytest/sample.js'
curl -i 'http://localhost:4747/file?path=sample.js'
```

### Set different base path and these files can no longer be accessed

```bash
export MINIFIERS_BASE_PATH=/tmp/differentdirectory
npm start
```

```bash
curl -i 'http://localhost:4747/file?path=/tmp/minifytest/sample.js'
curl -i 'http://localhost:4747/file?path=sample.js'
```

### Files outside of the basepath cannot be accessed

```bash
curl -i 'http://localhost:4747/file?path=../../../../../../../etc/passwd'
curl -i 'http://localhost:4747/file?path=/etc/passwd'
```

### CODE REVIEW

File: lib/validate-path.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+const defaultBasePath = path.dirname( __dirname );
+
+// Allow a path, but only if it falls under the base path after resolving.
+//
+// The base path can be set with the MINIFIERS_BASE_PATH environment variable.
+// If not set, it defaults to the minifiers directory.
+module.exports = ( userPath ) => {
+	const normalizedUserPath = path.normalize( userPath );
+
+	const basePath = path.resolve( process.env.MINIFIERS_BASE_PATH || defaultBasePath );
```
Comment (sgomes) [inline lib/validate-path.js]: Nitpick: we can move this to the top level of the module, so that it doesn't have to be recalculated every time (since we don't expect an environment variable to change at runtime).

File: README.md [hunk 2]
```diff
@@ -47,6 +47,10 @@ readme](https://www.npmjs.com/package/generic-pool).
 
 * Set `MINIFIERS_DISABLE_COMPRESSION` to 1 to disable compression.
 
+* Set `MINIFIERS_BASE_PATH` to choose a path of files served through the
+  /file endpoint. May be absolute or relative. If relative, relative to the
+  'lib' directory. Defaults to the repository root.
```
Comment (sgomes) [inline README.md]: > If relative, relative to the  'lib' directory.

In my testing, it appears to be relative to the shell working dir, not necessarily `lib`.

```sh
~/dev/minifiers %   MINIFIERS_BASE_PATH='.' node server.js

[...]
{ basePath: '/Users/sgomes/dev/minifiers' }
```

```sh
~/dev/minifiers/lib %   MINIFIERS_BASE_PATH='.' node ../server.js

[...]
{ basePath: '/Users/sgomes/dev/minifiers/lib' }
```

File: routes/file.js [hunk 3]
```diff
@@ -4,12 +4,25 @@ module.exports = async ( request, reply ) => {
...
 
-	const path = request.query.path;
-
 	let log = {};
+
+	const userPath = request.query.path;
+
+	// Check user's path falls in the base path.
+	const path = validatePath( userPath );
+	if ( path === false ) {
```
Comment (sgomes) [inline routes/file.js]: Nitpick: we can just do `!path` here, as we wouldn't want to do anything on any other falsy value anyway.

File: routes/file.js [hunk 4]
```diff
@@ -4,12 +4,25 @@ module.exports = async ( request, reply ) => {
...
+		// Error should look the same as a stat failed, in order to
+		// not leak information
+		const err = new Error( `ENOENT: no such file or directory, open '${ userPath }'` );
```
Comment (sgomes) [inline routes/file.js:20]: My first instinct would be to replace the system error with a custom message in the legitimate 404 scenario, by using a try/catch with a throw in the catch block, rather than attempting to duplicate the system message in the outside-the-base-path scenario. That seems safer in the long run, and less likely to leak should anything change.

I realise that it's effectively an API change, but I think we're early enough in the lifetime of the minifier that no-one is relying on specifically-worded error messages.

**Review by sgomes — COMMENTED** [review]
Thanks for these changes, @mreishus! 👍 I added some comments.

Also, we should probably add some tests to `file.test.js` to test these changes, to avoid future regressions.

File: README.md (hunk 2 above)
Comment (mreishus) [reply README.md]: Oh, right. I sent the default path up one directory after I realized the `tests/` directory needs to be readable, but didn't update the README.Fixed in 5fbf94c

File: lib/validate-path.js (hunk 1 above)
Comment (mreishus) [reply lib/validate-path.js]: :+1: Fixed in 006d671

File: routes/file.js (hunk 3 above)
Comment (mreishus) [reply routes/file.js]: Fixed in 08df020. I was initially scared of a really-edge-case of a file named "0", but upon checking, only php considers `"0"` to be false, javascript does not.

**Review by sgomes — APPROVED** [review]
LGTM! 👍 Thanks, @mreishus!

Feel free to leave the unit tests for a future PR, or to include them in this PR; whatever is most convenient.

**Comment by mreishus** [comment]
Working on the unit tests in #8. The unit tests currently require an external server to be spawned before they're run, but I want to make something that lets the tests spawn their own server, so I can try various settings of environment variables in the tests.
